### ISSUE TITLE
Memory allocation failed

### ISSUE THREAD

**Issue by mlomb** [aschey/stream-download-rs#36]
Hey! I'm trying to stream a big file ~31 GB of data, however my program is crashing.
I'm using the `BoundedStorageProvider` so I expect it it not allocate the full file at once, even if the content length is available (it is).

```rust
use std::error::Error;
use std::io::Read;
use std::num::NonZeroUsize;

use stream_download::storage::bounded::BoundedStorageProvider;
use stream_download::storage::memory::MemoryStorageProvider;
use stream_download::{Settings, StreamDownload};

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let reader = StreamDownload::new_http(
        "https://database.lichess.org/standard/lichess_db_standard_rated_2023-12.pgn.zst"
            .parse()?,
        BoundedStorageProvider::new(
            MemoryStorageProvider,
            NonZeroUsize::new(512 * 1024).unwrap(), // 512KB
        ),
        Settings::default(),
    )
    .await?;

    Ok(())
}
```

When running the code above, I get:
```
memory allocation of 31654288404 bytes failed
Aborted
```
(31654288404 is the size of the file, or close)

Cheers

**Comment by aschey** [comment]
Hey @mlomb, thanks for the bug report. When the content length is known, currently the `BoundedStorageProvider` prefers that over the provided buffer size. Since this provider was designed primarily for infinite streams, this wasn't something I put a lot of thought into.

For this scenario, I think it probably makes sense to compare the content length with the buffer size and use whichever one is smaller. That way we don't risk allocating extra space beyond the size of the content and using this to handle large files rather than infinite ones should work as expected.

I need to do some testing to make sure there's no inadvertent side effects from this change, but this should be a simple fix.

**Comment by aschey** [comment]
Also, out of curiosity, what's your use case here? This library was primarily designed with audio/video data in mind where you need to be able to play and seek through the content while it's still being downloaded. From the link you provided, I'm guessing that's a text-based database. Would be interested to hear how you intend to use it in case there's some specific use cases I hadn't considered.

**Comment by mlomb** [comment]
Oh I thought that was the `AdaptiveStorage` behaviour
In the `BoundedStorage` description it does not mention anything about content length so thought it did not matter

**Comment by mlomb** [comment]
> what's your use case here?

I need to parse almost all of Lichess database https://database.lichess.org (1.4TB), and for each game I have to do some processing and store only a bit of information in disk.
So I can just stream the download and process the games at the same time.

I was looking for a library that let me stream a URL (give me a `io::Read`) to pump the game parser (if you know another crate that does `URL → io::Read` tell me :P ). I'm using it in combination with [rust-pgn-reader](https://github.com/niklasf/rust-pgn-reader).

This is what I am using:

```rust
// raw data stream (may be compressed)
let raw_reader: Box<dyn io::Read> = if args.input.starts_with("http") {
    Box::new(
        StreamDownload::new_http(
            args.input.parse()?,
            BoundedStorageProvider::new(
                MemoryStorageProvider,
                NonZeroUsize::new(512 * 1024).unwrap(), // 512KB
            ),
            Settings::default(),
        )
        .await?,
    )
} else {
    Box::new(File::open(&args.input)?)
};

// decompress if necessary
let reader: Box<dyn io::Read> = if args.input.ends_with(".zst") {
    Box::new(zstd::Decoder::new(raw_reader)?)
} else {
    raw_reader
};

let mut game_reader = pgn_reader::BufferedReader::new(reader);

// ...
```

😄

**Comment by aschey** [comment]
> Oh I thought that was the `AdaptiveStorage` behaviour In the `BoundedStorage` description it does not mention anything about content length so thought it did not matter

Fair point, I should elaborate on this more in the docs. Technically you can make it work by using `BoundedStorage` with the buffer size equal to the content length and you would get the same behavior as `AdaptiveStorage`, but `BoundedStorage` has some overhead with all the calculations needed to make sure the buffer stays within bounds, so `AdaptiveStorage` bypasses that by automatically determining whether it needs to use the bounded buffer or not.

> I need to parse almost all of Lichess database https://database.lichess.org/ (1.65TB), and for each game I have to do some processing and store only a bit of information in disk.
So I can just stream the download and process the games at the same time.

Makes sense, I believe this should be able to handle that.

**Comment by aschey** [comment]
Fix released in `v0.4.2`. Thanks again!
